sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py

```diff
@@ -0,0 +1,491 @@
+    return reader
+
+  def writer(self):
+    self._timestamp = time.time()
+    writer = PatchedPubSubWriter(self, self._writer_class, (self.location,),
+                                 self._writer_kwargs)
+    return writer
+
+  @contextlib.contextmanager
+  def read_to_queue(self,
+                    seek_to_start=True,
+                    **kwargs):
+    self._assert_topic_exists()
+
+    reader_kwargs = self._reader_kwargs.copy()
+    reader_kwargs.update(kwargs)
+
+    if "subscription" in reader_kwargs:
+      created_subsciption = None
+    else:
+      created_subsciption = self._create_child_subscription(
+          seek_to_start=seek_to_start)
+      reader_kwargs["subscription"] = created_subsciption.name
+      assert created_subsciption in self._child_subscriptions
+
+    @functools.total_ordering
+    class PrioritizedTimestampedValue(TimestampedValue):
+
+      def __lt__(self, other):
+        return self.timestamp < other.timestamp
+
+    # Set arbitrary queue size limit to prevent OOM errors.
+    parsed_message_queue = queue.PriorityQueue(1000)
+    decoder = DecodeFromPubSub(
+        self.coder,
+        with_attributes=reader_kwargs.get("with_attributes", False),
+        timestamp_attribute=reader_kwargs["timestamp_attribute"])
+
+    def callback(msg):
+      msg.ack()
+      message = PubsubMessage._from_message(msg)
+      timestamped_value = next(decoder.process(message))
+      ordered_timestamped_value = PrioritizedTimestampedValue(
+          timestamped_value.value, timestamped_value.timestamp)
+      parsed_message_queue.put(ordered_timestamped_value)
+
+    sub_client = pubsub.SubscriberClient()
+    future = sub_client.subscribe(
+        reader_kwargs["subscription"], callback=callback)
+    try:
+      yield parsed_message_queue
+    finally:
+      future.cancel()
+      if created_subsciption is not None:
+        sub_client.delete_subscription(created_subsciption.name)
+        self._child_subscriptions.remove(created_subsciption)
+
+  def read(self, delay=0, timeout=5, **kwargs):
+    with self.read_to_queue(**kwargs) as message_queue:
+      time.sleep(delay)
+      while True:
+        try:
+          element = message_queue.get(timeout=timeout)
+          yield element
+        except queue.Empty:
+          return
+
+  def write(self, elements):
+    if self.requires_coder and self.coder is None:
+      # TODO(ostrokach): We might want to infer the element type from the first
+      # N elements, rather than reading the entire iterator.
+      elements = list(elements)
+      element_type = datatype_inference.infer_element_type(elements)
+      self.coder = coders.registry.get_coder(element_type)
+
+    writer_kwargs = self._writer_kwargs.copy()
+    attributes_fn = writer_kwargs.pop("attributes_fn", None)
+    # DirectRunner does not support timestamp_attribute
+    timestamp_attribute = writer_kwargs.pop("timestamp_attribute")
+
+    if writer_kwargs.get("with_attributes"):
```

Comment:
I think this code could be a lot simpler Pub/Sub message data was the element and the attributes contained any additional metadata used by the cache. I don't understand what `attributes_fn` is used for (and where is it set?).

Reply:
In order to be consistent with `WriteToPubSub` and `ReadFromPubSub`, *with_attributes* has to be a boolean flag indicating whether the input elements are bytes (*with_attributes=False*) or instances of `PubsubMessage` (*with_attributes = True*).

Forcing the user to manually convert their data into PubsubMessages is a bad user experience because, in the case of caching, the user probably just wants to examine the elements in the stream without applying any additional PTransforms. Therefore, I think *with_attributes* should be False in most cases. However, it is possible that the user wants to write their data to cache with a specific timestamp inferred from the data, rather than the event time of a given element. In order to make this possible for any data type, I added an `attributes_fn` argument, which takes a function that maps a given element to a timestamp.

If you think it would be simpler to get rid of this additional argument, please confirm and I will do so.

Reply:
I cleaned up the code around `PatchedPubSubReader` and `PatchedPubSubWriter` so hopefully it should be a bit easier to read.